File: definitions/savers/ScriptSaver.cs
using System.Collections.Generic;
using OSRSCache.script;

namespace OSRSCache.definitions.savers
{
	using ScriptDefinition = OSRSCache.definitions.ScriptDefinition;
	using OutputStream = OSRSCache.io.OutputStream;

	public class ScriptSaver
	{
		public virtual byte[] save(ScriptDefinition script)
		{
			int[] instructions = script.instructions;
			int[] intOperands = script.intOperands;
			string[] stringOperands = script.stringOperands;
			IDictionary<int, int>[] switches = script.switches;

			OutputStream @out = new OutputStream();
			@out.writeByte(0); // null string
			for (int i = 0; i < instructions.Length; ++i)
			{
				int opcode = instructions[i];
				@out.writeShort(opcode);
				if (opcode == (int) Opcodes.SCONST)
				{
					@out.writeString(stringOperands[i]);
				}
				else if (opcode < 100 && opcode != (int) Opcodes.RETURN && opcode != (int) Opcodes.POP_INT && opcode != (int) Opcodes.POP_STRING)
				{
					@out.writeInt(intOperands[i]);
				}
				else
				{
					@out.writeByte(intOperands[i]);
				}
			}
			@out.writeInt(instructions.Length);
			@out.writeShort(script.localIntCount);
			@out.writeShort(script.localStringCount);
			@out.writeShort(script.intStackCount);
			@out.writeShort(script.stringStackCount);
			int switchStart = @out.Offset;
			if (switches == null)
			{
				@out.writeByte(0);
			}
			else
			{
				@out.writeByte(switches.Length);
				foreach (IDictionary<int, int> s in switches)
				{
					@out.writeShort(s.Count);
					foreach (KeyValuePair<int, int> e in s.SetOfKeyValuePairs())
					{
						@out.writeInt(e.Key);
						@out.writeInt(e.Value);
					}
				}
			}
			int switchLength = @out.Offset - switchStart;
			@out.writeShort(switchLength);
			return @out.flip();
		}
	}

}

File: src/main/java/net/runelite/cache/definitions/savers/ScriptSaver.cpp
#include "ScriptSaver.h"

namespace net::runelite::cache::definitions::savers
{
	using Map = java::util::Map;
	using Entry = java::util::Map::Entry;
	using ScriptDefinition = net::runelite::cache::definitions::ScriptDefinition;
	using OutputStream = net::runelite::cache::io::OutputStream;
//	import static net.runelite.cache.script.Opcodes.SCONST;
//	import static net.runelite.cache.script.Opcodes.POP_INT;
//	import static net.runelite.cache.script.Opcodes.POP_STRING;
//	import static net.runelite.cache.script.Opcodes.RETURN;

	std::vector<signed char> ScriptSaver::save(std::shared_ptr<ScriptDefinition> script)
	{
		std::vector<int> instructions = script->getInstructions();
		std::vector<int> intOperands = script->getIntOperands();
		std::vector<std::wstring> stringOperands = script->getStringOperands();
		std::vector<std::unordered_map<int, int>> switches = script->getSwitches();

		std::shared_ptr<OutputStream> out = std::make_shared<OutputStream>();
		out->writeByte(0); // null string
		for (int i = 0; i < instructions.size(); ++i)
		{
			int opcode = instructions[i];
			out->writeShort(opcode);
			if (opcode == SCONST)
			{
				out->writeString(stringOperands[i]);
			}
			else if (opcode < 100 && opcode != RETURN && opcode != POP_INT && opcode != POP_STRING)
			{
				out->writeInt(intOperands[i]);
			}
			else
			{
				out->writeByte(intOperands[i]);
			}
		}
		out->writeInt(instructions.size());
		out->writeShort(script->getLocalIntCount());
		out->writeShort(script->getLocalStringCount());
		out->writeShort(script->getIntStackCount());
		out->writeShort(script->getStringStackCount());
		int switchStart = out->getOffset();
		if (switches.empty())
		{
			out->writeByte(0);
		}
		else
		{
			out->writeByte(switches.size());
			for (auto s : switches)
			{
				out->writeShort(s.size());
				for (auto e : s)
				{
					out->writeInt(e.first);
					out->writeInt(e.second);
				}
			}
		}
		int switchLength = out->getOffset() - switchStart;
		out->writeShort(switchLength);
		return out->flip();
	}
}
